tests/test_two_sided_market.py

```diff
@@ -0,0 +1,288 @@
+from math import isclose
+from unittest.mock import MagicMock
+
+import pytest
+from d3a.d3a_core.exceptions import BidNotFound, InvalidBid, InvalidBidOfferPair, InvalidTrade
+from d3a.events import MarketEvent
+from d3a.models.market import Bid, Offer
+from d3a.models.market.market_structures import TradeBidOfferInfo
+from d3a.models.market.two_sided import TwoSidedMarket
+from d3a.models.myco_matcher import PayAsBidMatcher, PayAsClearMatcher
+from d3a_interface.constants_limits import ConstSettings
+from pendulum import now
+
+
+@pytest.fixture
+def market():
+    return TwoSidedMarket(time_slot=now())
+
+
+@pytest.fixture
+def pac_market():
+    return PayAsClearMatcher()
+
+
+@pytest.fixture
+def market_matcher():
+    return PayAsBidMatcher()
+
+
+class TestTwoSidedMarket:
+    """Class Responsible for testing two sided market"s functionality."""
+
+    def test_double_sided_validate_requirements_satisfied(self, market):
+        offer = Offer("id", now(), 2, 2, "other", 2,
+                      requirements=[{"preferred_trading_partners": ["bid_id2"]}],
+                      attributes={"energy_type": "Green"})
+        bid = Bid("bid_id", now(), 9, 10, "B", 9,
+                  buyer_id="bid_id", requirements=[], attributes={})
+        with pytest.raises(InvalidBidOfferPair):
+            # should raise an exception as buyer_id is not in preferred_trading_partners
+            market.validate_requirements_satisfied(offer, bid)
+        bid.buyer_id = "bid_id2"
+        market.validate_requirements_satisfied(offer, bid)  # Should not raise any exceptions
+        bid.requirements.append({"energy_type": ["Grey"]})
+        with pytest.raises(InvalidBidOfferPair):
+            # should raise an exception as energy_type of offer needs to be in [Grey, ]
+            market.validate_requirements_satisfied(offer, bid)
+
+        # Adding another requirement that is satisfied, should not raise an exception
+        bid.requirements.append({"energy_type": ["Green"]})
+        market.validate_requirements_satisfied(offer, bid)
+
+        offer.requirements.append({"hashed_identity": "AreaX"})
+        # Above is an unsatisfied requirement but it will pass anw as
+        # there are other satisfied requirement
+        market.validate_requirements_satisfied(offer, bid)
+        bid.requirements = [{"hashed_identity": "AreaX"}]
+        with pytest.raises(InvalidBidOfferPair):
+            # should raise an exception as hashed_identity requirement of offer isn"t satisfied
+            market.validate_requirements_satisfied(offer, bid)
+        offer.attributes["hashed_identity"] = "AreaX"
+        market.validate_requirements_satisfied(offer, bid)
+
+    @pytest.mark.parametrize(
+        "bid_energy, offer_energy, clearing_rate, selected_energy", [
+            (2, 3, 2, 2.5),
+            (3, 2, 2, 2.5),
+            (2, 2, 2 / 2 + 1, 2),
```

Comment:
Any reason to have this calculation here?
```suggestion
            (2, 2, 2 , 2),
```

Reply:
I see that this is copy paste code, Sorry. But i guess you can change it anyways, right?